gd1_helpers/membership/gd1_model.py

```diff
@@ -155,17 +294,78 @@ def extra_ln_prior(self, params):
+        # cls.phi1_lim =(np.min(pawprint.skyprint['stream'].vertices[:,0]), np.max(pawprint.skyprint['stream'].vertices[:,0]))
+        cls.phi1_locs = get_grid(*cls.phi1_lim, cls.phi1_dens_step, pad_num=1).reshape(-1, 1)
+
+        cls.phi2_knots = get_grid(*cls.phi1_lim, knot_sep)  # knots every 10º
+
+        cls.pm1_knots = get_grid(*cls.phi1_lim, knot_sep)  # knots every 10º
+        cls.pm2_knots = get_grid(*cls.phi1_lim, knot_sep)  # knots every 10º
+
+        cls.variables = {
+            "phi1": GridGMMVariable(
+                param_priors={
+                    "zs": dist.Uniform(
+                        jnp.full(cls.phi1_locs.shape[0] - 1, -8),
+                        jnp.full(cls.phi1_locs.shape[0] - 1, 8),
+                    )
+                },
+                locs=cls.phi1_locs,
+                scales=np.full_like(cls.phi1_locs, cls.phi1_dens_step),
+                coord_bounds=cls.phi1_lim,
+            ),
+            "phi2": Normal1DSplineVariable(
+                param_priors={
+                    "mean": dist.Uniform(
+                        jnp.full_like(cls.phi2_knots, -4.0), jnp.full_like(cls.phi2_knots, 1.0)
+                    ),
+                    "ln_std": dist.Uniform(
+                        jnp.full_like(cls.phi2_knots, -2.0), jnp.full_like(cls.phi2_knots, 0.5)
+                    ),
+                },
+                knots=cls.phi2_knots,
+                coord_bounds=cls.coord_bounds["phi2"],
+            ),
+            "pm1": Normal1DSplineVariable(
+                param_priors={
+                    "mean": dist.Uniform(*cls.coord_bounds.get("pm1")).expand(
+                        cls.pm1_knots.shape
+                    ),
+                    "ln_std": dist.Uniform(-5, 0).expand(cls.pm1_knots.shape),  # ~20 km/s
+                },
+                knots=cls.pm1_knots,
+                coord_bounds=cls.coord_bounds.get("pm1"),
+            ),
+            "pm2": Normal1DSplineVariable(
+                param_priors={
+                    "mean": dist.Uniform(*cls.coord_bounds.get("pm2")).expand(cls.pm2_knots.shape),
+                    "ln_std": dist.Uniform(-5, 0).expand(cls.pm2_knots.shape),  # ~20 km/s
+                },
+                knots=cls.pm2_knots,
+                coord_bounds=cls.coord_bounds.get("pm2"),
+            ),
+        }
+        cls.data_required = {
+            "phi2": {"x": "phi1", "y": "phi2"},
+            "pm1": {"x": "phi1", "y": "pm1", "y_err": "pm1_err"},
+            "pm2": {"x": "phi1", "y": "pm2", "y_err": "pm2_err"},
+        }
+
+
+class OffTrackModel(Base, StreamModel):
     name = "offtrack"
 
     ln_N_dist = dist.Uniform(-5, 10)
 
-    dens_phi1_lim = (-60, 0)
-    dens_phi2_lim = (-3, 3)
+    dens_phi1_lim = (-100, 20)
+    dens_phi2_lim = (-8, 3.5)
 
-    dens_steps = [2.0, 0.25]
-    spar_steps = [8.0, 4.0]
+    dens_steps = np.array([3.0, 0.4]) # should find some optimal spacing here
+    spar_steps = 5*dens_steps
```

Comment:
In my original implementation, I had a "dense" grid of Gaussians centered on the spur, and a "sparse" grid elsewhere (dens and spar). But it looks like you now just have a dense grid covering the whole space. We can probably remove the "spar" parts then! But actually: I want to move to a model where we only have nodes inside of the stream pawprint (plus some buffer) to save on compute time...

Reply:
Oh I think I had misunderstood your original implementation. I thought you had a dense and sparse grid on everything to capture different scales. And agree on only using nodes in the stream pawprint, that's one of my goals for this week.